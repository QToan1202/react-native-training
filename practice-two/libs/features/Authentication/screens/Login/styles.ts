import { StyleSheet } from 'react-native'

import { COLORS } from '@practice-two/shared/constants'

const styles = StyleSheet.create({
  container: {
    flex: 1,
    paddingHorizontal: 32,
    backgroundColor: COLORS.PRIMARY,
    justifyContent: 'center',
  },
  input: {
    marginVertical: 35,
    rowGap: 20,
  },
})

export default styles
