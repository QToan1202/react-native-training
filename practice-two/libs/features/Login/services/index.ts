import { IUser, get } from '@practice-two/shared'

export const login = async (path: string, email: string, password: string): Promise<IUser> => {
  const users: IUser[] = await get(path, { params: { email } })

  if (!users.length) throw Error('Login fail, check email or password') // TODO: Extract to error message constant
  if (users.some((user: IUser) => user.password !== password))
    throw Error('Login fail, check email or password')

  return users[0]
}

export default login
