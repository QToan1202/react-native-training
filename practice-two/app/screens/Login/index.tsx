import { useCallback, useEffect } from 'react'
import { KeyboardAvoidingView, Platform } from 'react-native'
import { XStack, YStack } from 'tamagui'
import { SubmitHandler, useForm } from 'react-hook-form'
import { SafeAreaView } from 'react-native-safe-area-context'

import { NativeStackScreenProps } from '@react-navigation/native-stack'
import { RootStackParamList } from '@navigation/Stack'
import { Button, Heading, Input, Paragraph } from '@components'
import { IForm } from '@types'
import { COLORS, ERROR_MESSAGES } from '@constants'
import { useLogin } from '@hooks'
import { asyncStoreService } from '@services'
import { useAuthStore } from '@stores'

import styles from './styles'

export type LoginScreenProps = NativeStackScreenProps<RootStackParamList, 'Login'>

const Login = ({ navigation }: LoginScreenProps) => {
  const { mutate, data: user, isSuccess, status } = useLogin(process.env.USER_ENDPOINT)
  const loginFn = useAuthStore((state) => state.login)
  const { control, handleSubmit } = useForm<IForm>()
  const onSubmit: SubmitHandler<IForm> = useCallback(({ email, password }: IForm) => {
    mutate({
      email,
      password,
    })
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [])
  const handleSaveUser = useCallback(async () => {
    if (!isSuccess) return

    await asyncStoreService.save('user', user, () => loginFn(user))
  }, [isSuccess, loginFn, user])

  useEffect(() => {
    handleSaveUser()
  }, [handleSaveUser])

  const handleToSignUpScreen = useCallback(() => navigation.navigate('SignUp'), [navigation])
  return (
    <SafeAreaView style={styles.container}>
      <YStack space="$space.9" alignItems="center">
        <Heading content="Welcome to tradly" fontWeight="$2" letterSpacing="$2" />
        <Paragraph content="Login to your account" fontSize="$2" letterSpacing="$4" />
      </YStack>
      <KeyboardAvoidingView
        behavior={Platform.OS === 'ios' ? 'padding' : 'height'}
        style={styles.input}
      >
        <Input
          name="email"
          isShowError
          control={control}
          placeholder="Email/Mobile Number"
          placeholderTextColor={COLORS.WHITE}
          rules={{
            required: ERROR_MESSAGES.ACCOUNT.REQUIRED,
          }}
        />
        <Input
          name="password"
          isShowError
          secureTextEntry
          control={control}
          placeholder="Password"
          placeholderTextColor={COLORS.WHITE}
          rules={{
            required: ERROR_MESSAGES.PASSWORD.REQUIRED,
          }}
        />
      </KeyboardAvoidingView>
      <Button
        marginBottom="$space.6"
        title="login"
        variant="secondary"
        isDisable={status === 'pending'}
        onPress={handleSubmit(onSubmit)}
      />
      <YStack alignItems="center" space="$space.8">
        <Paragraph content="Forgot your password?" fontSize="$3" lineHeight="$4" />
        <XStack alignItems="center" columnGap="$space.1.5">
          <Paragraph content="Don't have an account?" fontSize="$3" lineHeight="$4" />
          <Button
            variant="tertiary"
            borderColor="$color.transparent"
            fontWeight="$3"
            fontSize="$3"
            lineHeight="$4"
            textTransform="none"
            title="Sign up"
            onPress={handleToSignUpScreen}
          />
        </XStack>
      </YStack>
    </SafeAreaView>
  )
}

export default Login
