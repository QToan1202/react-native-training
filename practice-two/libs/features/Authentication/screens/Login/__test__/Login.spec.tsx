import { render } from '@testing-library/react-native'

import Login from '../index'

describe('Testing Login screen', () => {
  it('Render screen successfully', () => {
    const component = render(<Login />).toJSON()

    expect(component).toMatchSnapshot()
  })
})
