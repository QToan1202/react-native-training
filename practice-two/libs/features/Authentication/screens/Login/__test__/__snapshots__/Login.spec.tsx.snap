// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing Login screen Render screen successfully 1`] = `
<RNCSafeAreaView
  edges={
    {
      "bottom": "additive",
      "left": "additive",
      "right": "additive",
      "top": "additive",
    }
  }
  style={
    {
      "backgroundColor": "#33907C",
      "flex": 1,
      "justifyContent": "center",
      "paddingHorizontal": 32,
    }
  }
>
  <View
    style={
      {
        "alignItems": "center",
        "rowGap": 50,
      }
    }
  >
    <Text
      style={
        [
          {
            "color": "#fff",
            "fontFamily": "Montserrat_700Bold",
            "fontSize": 24,
          },
          {
            "fontFamily": "Montserrat_500Medium",
          },
          undefined,
        ]
      }
    >
      Welcome to tradly
    </Text>
    <Text
      style={
        [
          {
            "color": "#fff",
            "fontFamily": "Montserrat_400Regular",
          },
          {
            "fontSize": 16,
          },
          undefined,
        ]
      }
    >
      Login to your account
    </Text>
  </View>
  <View
    onLayout={[Function]}
    style={
      [
        {
          "marginVertical": 35,
          "rowGap": 10,
        },
        {
          "paddingBottom": 0,
        },
      ]
    }
  >
    <View>
      <TextInput
        onBlur={[Function]}
        onChangeText={[Function]}
        onFocus={[Function]}
        placeholder="Email/Mobile Number"
        placeholderTextColor="#fff"
        style={
          [
            {
              "borderColor": "#fff",
              "borderRadius": 50,
              "borderWidth": 1,
              "color": "#fff",
              "fontFamily": "Montserrat_400Regular",
              "fontSize": 18,
              "paddingLeft": 15,
              "paddingVertical": 12,
            },
            {},
            undefined,
          ]
        }
        value=""
      />
      <View
        style={
          {
            "height": 20,
          }
        }
      />
    </View>
    <View>
      <TextInput
        onBlur={[Function]}
        onChangeText={[Function]}
        onFocus={[Function]}
        placeholder="Password"
        placeholderTextColor="#fff"
        secureTextEntry={true}
        style={
          [
            {
              "borderColor": "#fff",
              "borderRadius": 50,
              "borderWidth": 1,
              "color": "#fff",
              "fontFamily": "Montserrat_400Regular",
              "fontSize": 18,
              "paddingLeft": 15,
              "paddingVertical": 12,
            },
            {},
            undefined,
          ]
        }
        value=""
      />
      <View
        style={
          {
            "height": 20,
          }
        }
      />
    </View>
  </View>
  <View
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      {
        "backgroundColor": "#fff",
        "borderColor": "transparent",
        "borderRadius": 50,
        "borderWidth": 2,
        "marginBottom": 30,
        "opacity": 1,
        "paddingVertical": 13,
      }
    }
  >
    <View
      style={
        {
          "alignItems": "center",
          "columnGap": 6,
          "display": "flex",
          "flexDirection": "row",
          "justifyContent": "center",
        }
      }
    >
      <Text
        style={
          [
            {
              "color": "#fff",
              "fontFamily": "Montserrat_400Regular",
            },
            {
              "fontSize": 16,
            },
            [
              {
                "color": "#33907C",
              },
              {
                "fontFamily": "Montserrat_500Medium",
                "textAlign": "center",
                "textTransform": "capitalize",
              },
              undefined,
            ],
          ]
        }
      >
        login
      </Text>
    </View>
  </View>
  <View
    style={
      {
        "alignItems": "center",
        "rowGap": 45,
      }
    }
  >
    <Text
      style={
        [
          {
            "color": "#fff",
            "fontFamily": "Montserrat_400Regular",
          },
          {
            "fontSize": 18,
          },
          undefined,
        ]
      }
    >
      Forgot your password?
    </Text>
    <View
      style={
        {
          "alignItems": "center",
          "columnGap": 4,
          "flexDirection": "row",
        }
      }
    >
      <Text
        style={
          [
            {
              "color": "#fff",
              "fontFamily": "Montserrat_400Regular",
            },
            {
              "fontSize": 18,
            },
            undefined,
          ]
        }
      >
        Don't have an account?
      </Text>
      <View
        accessibilityState={
          {
            "busy": undefined,
            "checked": undefined,
            "disabled": undefined,
            "expanded": undefined,
            "selected": undefined,
          }
        }
        accessibilityValue={
          {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          {
            "backgroundColor": "transparent",
            "borderColor": "transparent",
            "borderRadius": 50,
            "borderWidth": 2,
            "opacity": 1,
            "paddingVertical": 13,
          }
        }
      >
        <View
          style={
            {
              "alignItems": "center",
              "columnGap": 6,
              "display": "flex",
              "flexDirection": "row",
              "justifyContent": "center",
            }
          }
        >
          <Text
            style={
              [
                {
                  "color": "#fff",
                  "fontFamily": "Montserrat_400Regular",
                },
                {
                  "fontSize": 16,
                },
                [
                  {
                    "color": "#fff",
                  },
                  {
                    "fontFamily": "Montserrat_500Medium",
                    "textAlign": "center",
                    "textTransform": "capitalize",
                  },
                  {
                    "fontFamily": "Montserrat_600SemiBold",
                    "fontSize": 18,
                    "textTransform": "none",
                  },
                ],
              ]
            }
          >
            Sign up
          </Text>
        </View>
      </View>
    </View>
  </View>
</RNCSafeAreaView>
`;
